File: Shared/Messages/AgentCommand.cs
using Kehlet.Functional;
// ReSharper disable ClassNeverInstantiated.Global

namespace Shared.Messages;

[AutoClosed(true)]
public partial record AgentCommand
{
    partial record ListProjectsCommand;

    partial record GetProjectDetailsCommand(string ProjectName);

    partial record ListProjectDirectoryCommand(string ProjectName, string Path);

    partial record OpenFileCommand(string ProjectName, string Path);

    partial record WriteFileCommand(string ProjectName, string FilePath, string Content, FileWriteMode Mode);

    partial record SectionReplaceCommand(string ProjectName, string FilePath, SectionIdentifiers SectionIdentifiers, string ReplacementContent);

    partial record TextReplaceCommand(string ProjectName, string FilePath, string Search, string Replacement);

    partial record RunDotnetTestCommand(string ProjectName);
}

File: Client/AgentCommand.fs
namespace Client

open System.Text.Json.Serialization
open FsCodec.SystemTextJson

type SectionIdentifiers = { Start: string; End: string }

[<JsonConverter(typeof<TypeSafeEnumConverter<FileWriteMode>>)>]
type FileWriteMode =
    | Append
    | Write

[<JsonConverter(typeof<UnionConverter<AgentCommand>>)>]
type AgentCommand =
    | ListProjectsCommand
    | GetProjectDetailsCommand of projectName: string
    | ListProjectDirectoryCommand of projectName: string * folderPath: string
    | OpenFileCommand of projectName: string * filePath: string
    | WriteFileCommand of projectName: string * filePath: string * content: string * fileWriteMode: FileWriteMode
    | TextReplaceSectionCommand of
        projectName: string *
        filePath: string *
        sectionIdentifiers: SectionIdentifiers *
        content: string
    | TextReplaceCommand of projectName: string * filePath: string * search: string * content: string
    | TextInsertBeforeCommand of projectName: string * filePath: string * search: string * content: string
    | TextInsertAfterCommand of projectName: string * filePath: string * search: string * content: string
    | LoadPageCommand of url: string
